File: association_rules/zig/ar.zig
const std = @import("std");

pub fn Itemset(comptime Items: []u32) type Itemset {
    return Itemset{ .items = Items };
}

pub fn FrequentItemset(Itemset: Itemset, Support: u32) type FrequentItemset {
    return FrequentItemset{ .itemset = Itemset, .support = Support };
}

pub fn isSubset(A: Itemset, B: Itemset) bool {
    var found = true;
    for (B.items) |item_b| {
        var itemFound = false;
        for (A.items) |item_a| {
            if (item_a == item_b) {
                itemFound = true;
                break;
            }
        }
        if (!itemFound) {
            found = false;
            break;
        }
    }
    return found;
}

pub fn isFrequent(Itemset: Itemset, Transactions: []Itemset, MinSupport: u32) bool {
    var count: u32 = 0;
    for (Transactions) |transaction| {
        if (isSubset(transaction, Itemset)) {
            count += 1;
        }
    }
    return count >= MinSupport;
}

pub fn generateFrequentItemsets(Transactions: []Itemset, NumItems: u32, MinSupport: u32) void {
    var itemsets: [0]Itemset = undefined;
    var itemsetsCount: u32 = 0;

    // Initialize with single items
    for (NumItems) |i| {
        itemsets[itemsetsCount] = Itemset(&[_]u32{i});
        itemsetsCount += 1;
    }

    // Generate frequent itemsets
    for (2..NumItems) |k| : (i) {
        var newItems: [0]Itemset = undefined;
        var newItemsCount: u32 = 0;

        for (0..itemsetsCount) |i| {
            for (i+1..itemsetsCount) |j| {
                // Create a candidate itemset
                var candidateItems: [k]u32 = undefined;
                for (0..k-1) |l| {
                    candidateItems[l] = itemsets[i].items[l];
                }
                candidateItems[k-1] = itemsets[j].items[k-1];
                const candidate = Itemset(candidateItems);

                // Check if the candidate itemset is frequent
                if (isFrequent(candidate, Transactions, MinSupport)) {
                    newItems[newItemsCount] = candidate;
                    newItemsCount += 1;
                }
            }
        }

        // Copy the new itemsets
        for (newItemsCount) |i| {
            itemsets[itemsetsCount] = newItems[i];
            itemsetsCount += 1;
        }
    }

    // Print frequent itemsets
    for (itemsetsCount) |i| {
        std.debug.print("Itemset: ");
        for (itemsets[i].items) |item| {
            std.debug.print("{d} ", .{item});
        }
        std.debug.println("");
    }
}

pub fn main() void {
    const numTransactions: u32 = 5;
    const numItems: u32 = 4;
    const minSupport: u32 = 2;

    // Create a sample dataset
    var transactions: [5]Itemset = undefined;
    transactions = &[
        Itemset(&[_]u32{1, 2, 3}),
        Itemset(&[_]u32{2, 3}),
        Itemset(&[_]u32{1, 3, 4}),
        Itemset(&[_]u32{1, 3}),
        Itemset(&[_]u32{1, 2, 3}),
    ];

    generateFrequentItemsets(transactions, numItems, minSupport);
}

File: association_rules/rust/ar.rs
struct Itemset {
    items: Vec<u32>,
}

struct FrequentItemset {
    itemset: Itemset,
    support: u32,
}

impl Itemset {
    fn new(items: Vec<u32>) -> Itemset {
        Itemset { items }
    }
}

impl FrequentItemset {
    fn new(itemset: Itemset, support: u32) -> FrequentItemset {
        FrequentItemset { itemset, support }
    }
}

fn is_subset(a: &Itemset, b: &Itemset) -> bool {
    b.items.iter().all(|item_b| a.items.contains(item_b))
}

fn is_frequent(itemset: &Itemset, transactions: &[Itemset], min_support: u32) -> bool {
    let count = transactions.iter().filter(|transaction| is_subset(transaction, itemset)).count();
    count as u32 >= min_support
}

fn generate_frequent_itemsets(transactions: &[Itemset], num_items: u32, min_support: u32) {
    let mut itemsets = Vec::new();

    // Initialize with single items
    for i in 1..=num_items {
        let itemset = Itemset::new(vec![i]);
        itemsets.push(itemset);
    }

    // Generate frequent itemsets
    for k in 2..=num_items {
        let mut new_items = Vec::new();

        for i in 0..itemsets.len() {
            for j in (i + 1)..itemsets.len() {
                // Create a candidate itemset
                let mut candidate_items = itemsets[i].items.clone();
                candidate_items.push(*itemsets[j].items.last().unwrap());
                let candidate = Itemset::new(candidate_items);

                // Check if the candidate itemset is frequent
                if is_frequent(&candidate, transactions, min_support) {
                    new_items.push(candidate);
                }
            }
        }

        // Copy the new itemsets
        itemsets.extend(new_items);
    }

    // Print frequent itemsets
    for itemset in itemsets {
        println!("Itemset: {:?}", itemset.items);
    }
}

fn main() {
    let num_transactions = 5;
    let num_items = 4;
    let min_support = 2;

    // Create a sample dataset
    let transactions = vec![
        Itemset::new(vec![1, 2, 3]),
        Itemset::new(vec![2, 3]),
        Itemset::new(vec![1, 3, 4]),
        Itemset::new(vec![1, 3]),
        Itemset::new(vec![1, 2, 3]),
    ];

    generate_frequent_itemsets(&transactions, num_items, min_support);
}
